opnode/rollup/driver/state_test.go

```diff
@@ -37,76 +37,183 @@ func (id testID) ID() eth.BlockID {
+	l2Head      eth.BlockID
+	l2Finalized eth.BlockID
+	l1Window    []eth.BlockID
+}
+
+type outputReturnArgs struct {
+	l2Head eth.BlockID
+	err    error
+}
+
+type stateTestCaseStep struct {
+	expectedL1Head testID
+	expectedL2Head testID
+	expectedWindow []testID
+
+	l1action func(t *testing.T, s *state, src *fakeChainSource, l1Heads chan eth.L1BlockRef)
+	l2action func(t *testing.T, expectedWindow []testID, s *state, src *fakeChainSource, outputIn chan outputArgs, outputReturn chan outputReturnArgs)
+	reorg    bool
 }
 
-type mockDriver struct {
-	mock.Mock
+func advanceL1(t *testing.T, s *state, src *fakeChainSource, l1Heads chan eth.L1BlockRef) {
+	l1Heads <- src.advanceL1()
 }
 
-func (m *mockDriver) requestEngineHead(ctx context.Context) (refL1 eth.BlockID, refL2 eth.BlockID, err error) {
-	returnArgs := m.Called(ctx)
-	refL1 = returnArgs.Get(0).(eth.BlockID)
-	refL2 = returnArgs.Get(1).(eth.BlockID)
-	err = returnArgs.Get(2).(error)
-	return
+func stutterL1(t *testing.T, s *state, src *fakeChainSource, l1Heads chan eth.L1BlockRef) {
+	l1Heads <- src.l1Head()
 }
 
-func (m *mockDriver) findSyncStart(ctx context.Context) (nextRefL1 eth.BlockID, refL2 eth.BlockID, err error) {
-	returnArgs := m.Called(ctx)
-	nextRefL1 = returnArgs.Get(0).(eth.BlockID)
-	refL2 = returnArgs.Get(1).(eth.BlockID)
-	err, _ = returnArgs.Get(2).(error)
-	return
+func stutterAdvance(t *testing.T, s *state, src *fakeChainSource, l1Heads chan eth.L1BlockRef) {
+	l1Heads <- src.l1Head()
+	l1Heads <- src.l1Head()
+	l1Heads <- src.l1Head()
+	l1Heads <- src.advanceL1()
+	l1Heads <- src.l1Head()
+	l1Heads <- src.l1Head()
+	l1Heads <- src.l1Head()
 }
 
-func (m *mockDriver) driverStep(ctx context.Context, nextRefL1 eth.BlockID, refL2 eth.BlockID, finalized eth.BlockID) (l2ID eth.BlockID, err error) {
-	returnArgs := m.Called(ctx, nextRefL1, refL2, finalized)
-	l2ID = returnArgs.Get(0).(eth.BlockID)
-	err, _ = returnArgs.Get(1).(error)
-	return
+func stutterL2(t *testing.T, expectedWindow []testID, s *state, src *fakeChainSource, outputIn chan outputArgs, outputReturn chan outputReturnArgs) {
+	select {
+	case <-outputIn:
+		t.Error("Got a step when no step should have occurred (l1 only advance)")
+	default:
+	}
+}
+
+func advanceL2(t *testing.T, expectedWindow []testID, s *state, src *fakeChainSource, outputIn chan outputArgs, outputReturn chan outputReturnArgs) {
+	args := <-outputIn
+	assert.Equal(t, int(s.Config.SeqWindowSize), len(args.l1Window), "Invalid L1 window size")
+	assert.Equal(t, len(expectedWindow), len(args.l1Window), "L1 Window size does not match expectedWindow")
+	for i := range expectedWindow {
+		assert.Equal(t, expectedWindow[i].ID(), args.l1Window[i], "Window elements must match")
+	}
+	outputReturn <- outputReturnArgs{l2Head: src.setL2Head(int(args.l2Head.Number) + 1).Self, err: nil}
+}
+
+func reorg__L2(t *testing.T, expectedWindow []testID, s *state, src *fakeChainSource, outputIn chan outputArgs, outputReturn chan outputReturnArgs) {
```

Comment:
maybe remove the underscores?

Reply:
this was to make the test cases below line up.